server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/DefaultMetaRepository.java

```diff
@@ -125,57 +152,124 @@ public CompletableFuture<Mirror> mirror(String id, Revision revision) {
                 throw new RepositoryMetadataException("failed to load the mirror configuration", e);
             }
 
-            final CompletableFuture<List<Credential>> credentials;
             if (Strings.isNullOrEmpty(c.credentialId())) {
-                credentials = projectCredentials();
+                if (!parent().repos().exists(repoName)) {
+                    throw mirrorNotFound(revision, mirrorFile);
+                }
+                return CompletableFuture.completedFuture(c.toMirror(parent(), Credential.FALLBACK));
+            }
+
+            final List<String> credentialRepoNames;
+            final boolean hasProjectCredential;
+            if (c.repoCredential()) {
```

Comment:
Suggestion) What do you think of adding a new resource repository layer to find a credential from a given ID? This logic could be simplified as we don't have to care if the ID is a repo credential.
```java
interface MetaResourceRepository {
    T find(String resourceName, Class<T> clazz);
}

Credential credential = repositoryRepository.find(mirror.credentialId(), Credential.class);
```

Reply:
I've used this logic not to fetch the resource one by one when multiple credentials are needed to be found. Let me encapsulate the logic into another class then.